### ISSUE TITLE
proccess is killed when using multiple running by the same seed.

### ISSUE THREAD

**Issue by czjghost** [pclucas14/mask-trick#1]
Thank you for you nice work, MIR and current paper.
When I try to run your "main.py" multiply under same seed, the process is often killed, then I turn the "pin_memory" off and adopt torch.cuda.empty_cache(), but it still occurs. Finally, I found the "register_buffer" leads in the process be killed. Could you give me some advice about how to release the tensors which have been register?
I am looking forward to your replay~Thank you in advance for your reply.

**Comment by czjghost** [comment]
What is more, I have try "gc.collect", "del buffer"..... but it still fails.

**Comment by pclucas14** [comment]
Hi, can you tell me which command exactly you are running ?

**Comment by czjghost** [comment]
python main.py --dataset cifar100 --method er_ace --mem_size 50 --supcon_temperature 0.2 --n_tasks 10 --task_free 1 --use_augs 1

python main.py --dataset cifar100 --method er_aml --mem_size 50 --supcon_temperature 0.2 --n_tasks 10  --task_free 1 --use_augs 1

![2384ac682ddbb3c7.png](https://user-images.githubusercontent.com/84562125/202485333-0abcdf22-19d6-44d3-9432-b96b21785905.png)

**Comment by pclucas14** [comment]
Can you share the full error message ? Also how much RAM does your computer have ?

**Comment by czjghost** [comment]
I try to add the loop and redefine the dataload which is similar to the implmentatiom in ASER and SCR

**Comment by pclucas14** [comment]
Can you share your modified code so that I can try to reproduce the error ?

**Comment by czjghost** [comment]
By the way, I delete the "if n_seen > args.samples_per_task > 0: break". Does it cause this case?

**Comment by pclucas14** [comment]
I have never encountered the error you are describing. If you would like me to help, then I need to run your code to reproduce the error :p. Otherwise I can't help you. Maybe try setting `num_workers=0` but that's all the advice I have

**Comment by czjghost** [comment]
> Can you share the full error message ? Also how much RAM does your computer have ?
> 

I donot know the ARM how much my service has. But I can run 3 times continually on cifar10, others are less.

**Comment by czjghost** [comment]
[aml.zip](https://github.com/pclucas14/AML/files/10033003/aml.zip)

**Comment by pclucas14** [comment]
ok thanks. Going to look at this in hopefully within a few days. In the meantime, if this is pressing please try setting the number of workers to 0 in the dataloader. I will keep you posted

**Comment by czjghost** [comment]
> I have never encountered the error you are describing. If you would like me to help, then I need to run your code to reproduce the error :p. Otherwise I can't help you. Maybe try setting `num_workers=0` but that's all the advice I have

I set num_workers=0 all the time. And I run this code in my laptop, when running several times continuall, my “task manager” shows my Memory(2 * 16GB ddr4 3200MHz) is full but the gpu memory(RTX 3070 8GB) is still enough to running.

**Comment by czjghost** [comment]
> ok thanks. Going to look at this in hopefully within a few days. In the meantime, if this is pressing please try setting the number of workers to 0 in the dataloader. I will keep you posted

Okay, thank you for your reply.  I do apologize for occupying so much of your time.

**Comment by pclucas14** [comment]
If your process is being killed then you are running out of RAM, not of GPU memory. I would suggest to run this on google collab instead, you will have enough ram there

**Comment by czjghost** [comment]
> If your process is being killed then you are running out of RAM, not of GPU memory. I would suggest to run this on google collab instead, you will have enough ram there

OK, thanks, by the way, cound I abandon the tensors of "register_buffer" and set them as common tensor? then maybe I cound running it?

**Comment by pclucas14** [comment]
This should not change anything. All this does is to make sure that they will change device (go to gpu) when you do `buffer.gpu()`

**Comment by czjghost** [comment]
> This should not change anything. All this does is to make sure that they will change device (go to gpu) when you do `buffer.gpu()`
> 

emmm..okay, I try to test it google collab

**Comment by czjghost** [comment]
Excuse me, I run the code in Google colab, but .... I get the same result as running in my service.
https://colab.research.google.com/drive/1rgtrTxky-DwbPegIdHxz8BuBfeBa5vZi

the txt result and running logger, I save at below:
https://drive.google.com/drive/folders/1BOSzkjgcExY6goyBoEQoBaMDSnxfgVZT?usp=sharing

In my opinion, it shows that when implement the "register_buffer" several times, the memory cannot offer enough space to allocate, maybe if there exists a way to erase the space of current running that has registered in buffer, this problem will be tackled.

Hope you cound help me when you have spare time. Thank you for your reply and your time in advance.

**Comment by pclucas14** [comment]
hi, I have requested access to the google collab, can you grant me access so that I can have a look ?

**Comment by czjghost** [comment]
> hi, I have requested access to the google collab, can you grant me access so that I can have a look ?

sorry about that, I have abandon it, the file is gone...  the result is below. the code is the same as I send to you except delete the "break" in "if i % 50 ....."

I have shared  the google collab.
https://colab.research.google.com/drive/1rgtrTxky-DwbPegIdHxz8BuBfeBa5vZi?usp=sharing

**Comment by czjghost** [comment]
[result-20221121T111326Z-001.zip](https://github.com/pclucas14/AML/files/10055973/result-20221121T111326Z-001.zip)

**Comment by czjghost** [comment]
the main problem is below, it shows that system is unable to allocate memory to current running and abort current program and make another program run.

Task #0 --> Train Classifier

0 / 1000
tcmalloc: large alloc 12288000000 bytes == 0x7fbd7d940000 @  0x7fc202c64b6b 0x7fc202c84379 0x7fc18bab7d57 0x7fc18baa5bc3 0x7fc1b59b06af 0x7fc1b59b1020 0x7fc1b59b1074 0x7fc1b5ec2def 0x7fc1b6734b8b 0x7fc1b64801d3 0x7fc1b670fb8f 0x7fc1b64bdbf7 0x7fc1dda13860 0x7fc1b5eca1a4 0x7fc1b6984d15 0x7fc1b622a80e 0x7fc1b670e375 0x7fc1b62701b2 0x7fc1dd6f3ce7 0x58f795 0x510bf2 0x58fd37 0x510325 0x58fd37 0x510325 0x58fd37 0x510325 0x58fd37 0x50ca37 0x5b4ee6 0x6005a3
^C
Files already downloaded and verified
Files already downloaded and verified

Task #0 --> Train Classifier

0 / 1000
created buffer x	 torch.float32, torch.Size([3, 32, 32])
created buffer y	 torch.int64, torch.Size([])
created buffer t	 torch.int64, torch.Size([])

**Comment by pclucas14** [comment]
Hi, 

I tried the following two commands (that you posted above) with your downloaded code 

```
python main.py --dataset cifar100 --method er_ace --mem_size 50 --supcon_temperature 0.2 --n_tasks 10 --task_free 1 --use_augs 1

python main.py --dataset cifar100 --method er_aml --mem_size 50 --supcon_temperature 0.2 --n_tasks 10 --task_free 1 --use_augs 1
```

And it seems to work fine. Before modifying this codebase, was the code running on your computer ?

**Comment by pclucas14** [comment]
Ok I was able to reproduce the error. Can you try and deleting the model and agent before starting a new run ? 

`del model; del agent`
at the end of your loop (before starting a new run)

**Comment by pclucas14** [comment]
Ok I found the issue : The following line 
```
#    logger = Logger(args, holder)
    args.mem_size = args.mem_size * args.n_classes
```

gets executed every time you do a new run, so `args.mem_size` keeps growing each run. After a few runs, when you are creating the buffer, you are asking for `args.mem_size * args.n_classes ** run_number` which is too big. 

If you but the line I copy pasted **outside your run loop** you should be fine

**Comment by czjghost** [comment]
Oh, no... I make a bad mistake... sorry for occupying so much of your time..

**Comment by czjghost** [comment]
thank you very much, it works now, sorry for my negligence when changing single running code into multiply runnings code.

**Comment by pclucas14** [comment]
no problem, good luck!
